enterprise/internal/campaigns/syncer.go

```diff
@@ -13,37 +14,226 @@ import (
+		priorityNotify: make(chan []int64, 500),
+		syncers:        make(map[int64]*ChangesetSyncer),
+	}
+
+	services, err := repoStore.ListExternalServices(ctx, repos.StoreListExternalServicesArgs{})
+	if err != nil {
+		log15.Error("Fetching initial external services", "err", err)
+	}
+
+	// Add and start syncers
+	for _, service := range services {
+		r.Add(service.ID)
+	}
+
+	go r.handlePriorityItems()
+
+	return r
+}
+
+// Add adds a syncer for the supplied external service if it hasn't already been added and starts it.
+func (s *SyncRegistry) Add(extServiceID int64) {
+	ctx, cancel := context.WithTimeout(s.Parent, 10*time.Second)
+	defer cancel()
+	services, err := s.ReposStore.ListExternalServices(ctx, repos.StoreListExternalServicesArgs{
+		IDs: []int64{extServiceID},
+	})
+	if err != nil {
+		log15.Error("Listing external services", "err", err)
+		return
+	}
+	if len(services) < 1 {
+		return
+	}
+
+	service := services[0]
+
+	switch service.Kind {
+	case "GITHUB", "BITBUCKETSERVER":
+	// Supported by campaigns
+	default:
+		log15.Warn("Syncer not started for unsupported code host", "kind", service.Kind)
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.syncers[extServiceID]; ok {
+		// Already added
+		return
+	}
+
+	// We need to be able to cancel the syncer if the service is removed
+	ctx, cancel = context.WithCancel(s.Parent)
+
+	syncer := &ChangesetSyncer{
+		Store:             s.Store,
+		ReposStore:        s.ReposStore,
+		HTTPFactory:       s.HTTPFactory,
+		externalServiceID: extServiceID,
+		cancel:            cancel,
+	}
+
+	go syncer.Run(ctx)
+}
+
+// handlePriorityItems fetches changesets in the priority queue from the db and passes them
+// to the appropriate syncer
+func (s *SyncRegistry) handlePriorityItems() {
+	fetchSyncData := func(ids []int64) ([]campaigns.ChangesetSyncData, error) {
+		ctx, cancel := context.WithTimeout(s.Parent, 10*time.Second)
+		defer cancel()
+		return s.Store.ListChangesetSyncData(ctx, ListChangesetSyncDataOpts{ChangesetIDs: ids})
+	}
+	for {
+		select {
+		case <-s.Parent.Done():
+			return
+		case ids := <-s.priorityNotify:
+			syncData, err := fetchSyncData(ids)
+			if err != nil {
+				log15.Error("Fetching sync data", "err", err)
+				continue
+			}
+
+			// Gather all external services requested
+			changesetsByService := make(map[int64][]int64)
+			for _, d := range syncData {
+				for _, id := range d.ExternalServiceIDs {
+					if _, ok := changesetsByService[id]; !ok {
+						changesetsByService[id] = make([]int64, 0)
+					}
+				}
+			}
+
+			// Spread ids among syncers
+			for _, sd := range syncData {
+				externalService := shardChangeset(sd.ChangesetID, sd.ExternalServiceIDs)
+				changesetsByService[externalService] = append(changesetsByService[externalService], sd.ChangesetID)
+			}
+
+			// Anonymous func so we can use defer
+			func() {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				for svcID, changesets := range changesetsByService {
+					syncer, ok := s.syncers[svcID]
```

Comment:
What about just guard this single line?
```suggestion
					s.mu.Lock()
					syncer, ok := s.syncers[svcID]
					s.mu.Unlock()
```

Reply:
I don't think it makes sense to keep grabbing and releasing a lock inside a loop. It also make it harder to defer the unlock.

The overall function shouldn't block for long so I think it's safe

Reply:
> I don't think it makes sense to keep grabbing and releasing a lock inside a loop.

Why?